junit-jupiter-engine/src/test/java/org/junit/jupiter/engine/extension/TestInstancePostProcessorAndPreDestroyCallbackTests.java

```diff
@@ -0,0 +1,240 @@
+		// @formatter:off
+        assertBeforeAllAndAfterAllCallbacks(ThirdLevelTestCase.class,
+			"fooPostProcessTestInstance",
+			"barPostProcessTestInstance",
+				"bazPostProcessTestInstance",
+					"quuxPostProcessTestInstance",
+						"test-3",
+					"quuxPreDestroyTestInstance",
+				"bazPreDestroyTestInstance",
+			"barPreDestroyTestInstance",
+			"fooPreDestroyTestInstance"
+        );
+        // @formatter:on
+	}
+
+	@Test
+	void preDestroyTestInstanceMethodThrowsAnException() {
+		// @formatter:off
+        assertBeforeAllAndAfterAllCallbacks(ExceptionInTestInstancePreDestroyCallbackTestCase.class, 0,
+			"exceptionPostProcessTestInstance",
+				"test",
+			"exceptionThrowingTestInstancePreDestroyCallback"
+        );
+        // @formatter:on
+	}
+
+	@Test
+	void postProcessTestInstanceMethodThrowsAnException() {
+		// @formatter:off
+        assertBeforeAllAndAfterAllCallbacks(ExceptionInTestInstancePostProcessorTestCase.class, 0,
+			"exceptionThrowingTestInstancePostProcessor"
+        );
+        // @formatter:on
+	}
+
+	private void assertBeforeAllAndAfterAllCallbacks(Class<?> testClass, String... expectedCalls) {
+		assertBeforeAllAndAfterAllCallbacks(testClass, 1, expectedCalls);
+	}
+
+	private void assertBeforeAllAndAfterAllCallbacks(Class<?> testClass, int testsSuccessful, String... expectedCalls) {
+
+		callSequence.clear();
+
+		executeTestsForClass(testClass).testEvents()//
+				.assertStatistics(stats -> stats.started(1).succeeded(testsSuccessful));
+
+		assertEquals(asList(expectedCalls), callSequence, () -> "wrong call sequence for " + testClass.getName());
+	}
+
+	// -------------------------------------------------------------------------
+
+	// Must NOT be private; otherwise, the @Test method gets discovered but never executed.
+	@ExtendWith({ FooClassLevelCallbacks.class, BarClassLevelCallbacks.class })
+	static class TopLevelTestCase {
+
+		@Test
+		void test() {
+			callSequence.add("test-1");
+		}
+	}
+
+	// Must NOT be private; otherwise, the @Test method gets discovered but never executed.
+	@ExtendWith(BazClassLevelCallbacks.class)
+	static class SecondLevelTestCase extends TopLevelTestCase {
+
+		@Test
+		@Override
+		void test() {
+			callSequence.add("test-2");
+		}
+	}
+
+	@ExtendWith(QuuxClassLevelCallbacks.class)
+	static class ThirdLevelTestCase extends SecondLevelTestCase {
+
+		@Test
+		@Override
+		void test() {
+			callSequence.add("test-3");
+		}
+	}
+
+	@ExtendWith(ExceptionThrowingTestInstancePreDestroyCallback.class)
+	static class ExceptionInTestInstancePreDestroyCallbackTestCase {
+		@Test
+		void test() {
+			callSequence.add("test");
+		}
+	}
+
+	@ExtendWith(ExceptionThrowingTestInstancePostProcessor.class)
+	static class ExceptionInTestInstancePostProcessorTestCase {
+		@Test
+		void test() {
+			callSequence.add("test");
+		}
+	}
+
+	// -------------------------------------------------------------------------
+
+	static class FooClassLevelCallbacks extends AbstractClassLevelCallbacks {
```

Comment:
I think these `...ClassLevelCallbacks` names are refering to the beforeAll/afterAll callbacks, maybe we should find a name that fits this initialization/destroy use case instead

Reply:
Maybe just `...TestInstanceCallbacks`

Reply:
I thought it was named this way as defined via as @ExtendWith at the class level, agree @BeforeAll associated with the class/static, I'll follow your suggestion to make it more appropriate to what I'm testing.